packages/accounts-controller/src/AccountsController.ts

```diff
@@ -1160,6 +1126,36 @@ export class AccountsController extends BaseController<
     return accountsState;
   }
 
+  /**
+   * Handles the change in multichain network by updating the selected account.
+   *
+   * @param id - The EVM client ID or non-EVM chain ID that changed.
+   */
+  readonly #handleMultichainNetworkChange = (
+    id: NetworkClientId | CaipChainId,
+  ) => {
+    let accountId: string;
+
+    // We only support non-EVM Caip chain IDs at the moment. Ex Solana and Bitcoin
+    // MultichainNetworkController will handle throwing an error if the Caip chain ID is not supported
+    if (isCaipChainId(id)) {
+      // Update selected account to non evm account
+      const lastSelectedNonEvmAccount = this.getSelectedMultichainAccount(id);
+      // @ts-expect-error - This should never be undefined, otherwise it's a bug that should be handled
+      accountId = lastSelectedNonEvmAccount.id;
+    } else {
+      // Update selected account to evm account
+      const lastSelectedEvmAccount = this.getSelectedAccount();
+      accountId = lastSelectedEvmAccount.id;
+    }
+
+    this.update((currentState) => {
+      currentState.internalAccounts.accounts[accountId].metadata.lastSelected =
+        Date.now();
+      currentState.internalAccounts.selectedAccount = accountId;
+    });
+  };
```

Comment:
I assume we're not calling `setSelectedAccount` here to avoid triggering the "account change" event?

If yes, we should add a comment to explain this, cause this could easily be refactored and since there's no comment, we could make a mistake here.

Reply:
Also, I think that changing the account without publishing this event might have some side effects (like for other controllers that listen to this change event).

I'm not sure how to avoid infinite loop with those inter-dependent events, but that's something we might wanna re-visit at some point.

Reply:
Updated with comment. Yeah it's a tricky problem. We're just being extra careful not to subscribe and publish to the same controller in each handler atm. Regarding side effects, if another controller relies on account changes, it should subscribe to the state change event and manage its own logic